### ISSUE TITLE
Add animbrush functionality + sprite sheets

### ISSUE THREAD

**Issue by mriale** [mriale/PyDPainter#122]
- Grab animation frames as a brush
- Load and save animbrushes
- grab sprite sheets as animbrushes
- generate sprite sheets from animbrushes

**Comment by Sleuthgame** [comment]
Ambitious !

I'm trying to find some examples.

**Comment by mriale** [comment]
Many good and bad examples here:
https://www.spriters-resource.com/

**Comment by Sleuthgame** [comment]
[Some animbrush as animated icons here ](https://aminet.net/search?query=animbrush)
Not very funny.

**Comment by mriale** [comment]
Also add support for fonts in sprite sheet format. This should work with the text tool.

**Comment by Sleuthgame** [comment]
Lot of fonts in sprite sheet here : https://github.com/ianhan/BitmapFonts.

**Comment by mriale** [comment]
Some bugs still:
- drawmode cycle with animbrush doesn't change colors
- scaling while line drawing with animbrush doesn't work consistently

**Comment by Sleuthgame** [comment]
Working. I superpose 4 sprites of the evil in left in an animation and grab it as a 4 frames animbrush and was able to use it (on right).
![barb2-test1](https://github.com/user-attachments/assets/1f8e8828-9ad6-4d48-84ea-5ee59406eaaa)
Curiously GIF export is messing with my current tool cursor.

**Comment by Sleuthgame** [comment]
I did a test on a dancing animation (268 frames) I was working on. OMG it's so great fun ! Can't wait for the cycle mode support. (In reality, I can wait, no stress on you).

**Comment by mriale** [comment]
Color cycling and scaling for animbrushes should now work.

**Comment by Sleuthgame** [comment]
Hi !
If I already have **Brush** tool selected. ALT-B or menu switch to anim brush pick-up but in **Polygon mode**.

**Comment by Sleuthgame** [comment]
If I swap or change the color of my animbrush (BG->FG) to prepare for cycle, it affects only the current frame of the animbrush.

**Comment by Sleuthgame** [comment]
Scaling is working fine (268 frame animbrush). It's a little tricky to hit +/- while using Windows key and clicking, but I don't have any UX suggestion. It's already a lot of fun.

**Comment by mriale** [comment]
> If I swap or change the color of my animbrush (BG->FG) to prepare for cycle, it affects only the current frame of the animbrush.

Yes, I haven't fixed the Brush menu yet to propagate changes to all animbrush frames.

**Comment by mriale** [comment]
The brush menu items should work with all the frames of animbrushes now. It also brings up a progress bar if the brush operations are taking too long like for a big brush with a lot of frames.

Also I noticed that alt-grabbing an amimbrush wasn't implemented so I added that as well.

**Comment by Sleuthgame** [comment]
Sorry. I'm experiencing crashes while preparing sprites sheets into animation to grab some animbrushs. I'm try to figure out why.

**Comment by Sleuthgame** [comment]
Wanna dance ? 268 frames animbrush, baby ! Yeah ! (4 colors, PAL low-res)

![Dancing-girls](https://github.com/user-attachments/assets/1edacacd-08f9-4164-b951-5e295747cc4b)

It's a wanabe tutorial. I want to make another with a walking man loop.

**Comment by mriale** [comment]
> Sorry. I'm experiencing crashes while preparing sprites sheets into animation to grab some animbrushs. (from the barbarian 2 sprites, IDKW)
> 
> I'm try to figure out why. Is there some debug mode I can use to trace what's happening ?

When PyDPainter crashes, there is a file `stackdump.txt` written out in your home directory under the `.pydpainter-recover` directory. In Windows the home directory is C:\Users\ *username*. That would help me see where it is crashing.

**Comment by Sleuthgame** [comment]
Good to know. 👍 

Last one :
[stackdump.txt](https://github.com/user-attachments/files/17185427/stackdump.txt)
![recover](https://github.com/user-attachments/assets/b0eefa78-59b7-4373-b123-a293a5d92643)

**Comment by Sleuthgame** [comment]
Previous one. Same picture.
![recover](https://github.com/user-attachments/assets/42d9d5c2-870e-4840-b704-59a0d38c5f1c)
[stackdump.txt](https://github.com/user-attachments/files/17185440/stackdump.txt)

**Comment by mriale** [comment]
@Sleuthgame I think I fixed the 2 bugs from the above stack dumps. I couldn't recreate them, so I can't verify that I actually fixed them. Please take a look and let me know.

**Comment by Sleuthgame** [comment]
Good ! I'll try the same process and let you know.
thanks !
Stephane

**Comment by mriale** [comment]
@Sleuthgame Also just fixed another weird crash that happened if you drag out a brush while holding the `b` key down.

**Comment by mriale** [comment]
@Sleuthgame And in fixing the crash, it make polygonal brushes quit working. Hopefully that's fixed now too.

**Comment by mriale** [comment]
I'm playing around with a user interface for grabbing a simple grid of brushes. It's currently in `develop`. To trigger it:
* click on the brush tool first with the left button then right-click on it
* this now brings up the Brush Grid requestor
* If you click on "Visual" it brings up a grid you can drag out
  * while dragging, type `x` to increase the number of cells in the X direction or `shift-x` to decrease the number of cells
  * the `y` key does the same thing for the Y direction
* the OK button doesn't actually do anything yet

I'm not sure I like the single-drag user interface. It might be better to be able to adjust it multiple times with handles or just grabbing the grid squares. The `x` and `y` keys are not discoverable either.

I am thinking I'll add more to the requestor to be able to grab sprite sheets that are not uniform. Maybe buttons to detect white space in X and Y?

**Comment by Sleuthgame** [comment]
I think it's a good start. Obviously, a real visible icon (brush with an italic A in it ? Videogame sprite ?) would be easier.

I think a good real-life test is to check that it works with sprite sheets like the ones on this site.

https://www.spriters-resource.com/amiga_amiga_cd32/barbarianiiaxeofrage/sheet/128811/

**Comment by Sleuthgame** [comment]
I have a look. What are the justify and arrows buttons ?
It will be efficient to add the GRAB animbrush button directly here.

**Comment by mriale** [comment]
I changed Justify to Align and added spinners.

My idea for the Strict: No option is make the grid be flexible in the X or Y direction and look for whitespace to decide what size to make that frame of the animbrush. The Align options determine which way to align animbrush frames that are smaller than the grid size.

The idea is to make the requestor draggable like the palette and stencil requestors. The grid will be shown underneath and be able to be dragged directly or by changing the numbers.

This is still a concept of a GUI in progress.

**Comment by Sleuthgame** [comment]
Looks like you know where you're going. Instead, I worked with evenly spaced sprite boards on a 32x32px or 16x16 grid for some AMOS game. So I'm not in the best position to evaluate this method.

Aren't you planning on releasing the "classic" Animbrush functions first? All that's missing is the Save/Load and some polish?

**Comment by mriale** [comment]
So here is the user interface that I've been working on with a flexible grid to grab frames:
![image](https://github.com/user-attachments/assets/abe2b87e-4ccc-46a2-8985-fd7083f00719)

I think it doesn't work well, though, and I'm planning to change it.

I'm planning to convert a brush with sequence of images into an animbrush. A new requestor will show some of the same grid settings and show a frame at a time. Not sure I'll have much time to work on it in the next week, though.

**Comment by Sleuthgame** [comment]
OK, but add a **Grab** button right here between Ok and Cancel. The Animbrush workflow is quite forgotten for most people.

I am on a full time job too now, but I will have REAL PAID holidays.

**Comment by mriale** [comment]
I added load and save animbrush in IFF and GIF formats. The IFF animbrushes even transfer back and forth from DPaint III and V. The GIF load and save also works as you'd expect.

Next is the animbrush Settings requester with ping-pong and such.

**Comment by Sleuthgame** [comment]
OK, fine. The end of this is near. You tell me when you fix your process/UI for this, so I can write a doc / tuto for this new feature.

**Comment by mriale** [comment]
DPaint III animbrush settings for reference:
![image](https://github.com/user-attachments/assets/dbc5aeb6-b757-4801-a27d-ffd4f0d30a13)

**Comment by Sleuthgame** [comment]
Some test I made using a custom animbrush as painting brush :
- rotating brush : nice random brush fx
- glowing star
![image](https://github.com/user-attachments/assets/385f0c21-c077-4191-9c09-1c1c291c5e5d)

**Comment by Sleuthgame** [comment]
If I trigger **Animbrush Pick Up** with the current **Brush** tool selected, it fails and activate Polygonal brush instead.

**Comment by Sleuthgame** [comment]
That's nice ! Maybe a little too fast for Dot tools. I had to trace very quickly with my mouse.
![image](https://github.com/user-attachments/assets/a8f1da43-3e7b-409f-ba51-e77f909a2aa0)

**Comment by Sleuthgame** [comment]
Interesting effet. The starting frame of the animbrush change it all. 6 frames Animbrush, 6 different pens (Draw tool)
![image](https://github.com/user-attachments/assets/c7eb0278-6f4a-44b5-89da-8fb91f5fee01)

**Comment by Sleuthgame** [comment]
Not sure if **Continuous** drawing mode should behave like this. I think it should shuffle the animbrush frames.
First line : **N Total** 6. Next, **Continuous**. Same Animbrush.
7 and 8 won't change the frame ? Or it is but is all staked when drawing ?
![image](https://github.com/user-attachments/assets/20432e5e-94e6-461b-b962-e156158f48ea)

**Comment by Sleuthgame** [comment]
My Star and Circles dots animbrushes for testing.
[ANIMBRUSH.zip](https://github.com/user-attachments/files/18180177/ANIMBRUSH.zip)

**Comment by Sleuthgame** [comment]
> ![image](https://private-user-images.githubusercontent.com/26576606/382441087-abe2b87e-4ccc-46a2-8985-fd7083f00719.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzUyOTUxNzgsIm5iZiI6MTczNTI5NDg3OCwicGF0aCI6Ii8yNjU3NjYwNi8zODI0NDEwODctYWJlMmI4N2UtNGNjYy00NmEyLTg5ODUtZmQ3MDgzZjAwNzE5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjI3VDEwMjExOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTkxYzUxZTAxZWJjN2ExNjRiOWVkZjY0MWY3OWEyM2RhYzg1MzI4NTkxODZjMGEyZGU4N2U3ZTgxZjJlNjEzZWEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.5VZ2edzg6SK890GmFfl_oIcQK3Vsid8BH0gZaFRbIQk)

> I'm planning to convert a brush with sequence of images into an animbrush. A new requestor will show some of the same grid settings and show a frame at a time. Not sure I'll have much time to work on it in the next week, though.

Today, I take the time to figure out how it's work. It's terrific ! I manage to grab an entire demoscene font as a animbrush.

My big misstatement is that I expect Animbrush options to trigger on brush icon with right-clcik after animbrush grabbing. False. OK. So it's a brush option. Maybe just add a new entrance in the animbrush menu, is more appealing and easy to find.

Instead of OK, rename the button as GRAB Now ! Or something obvious.

A **pause** animbrush with 9 or in the menu will be usefull, especially with font or large sprite. Sometime, you don't want to animate and the whole animbrush is just a convenient multiple brushs browser.
IE : I grab a demoscene colored font. I "write some text", pasting brush with it browsing through all letters and numbers as frame. Very efficient.

**Comment by Sleuthgame** [comment]
I'm sorry. I try hard... but... I CAN'T BREAK IT ! 😆 
😄 🏆 
Congratulations to you!

**Comment by Sleuthgame** [comment]
I save my **Pacman.anbr** animbrush and when loading it later it mess/decay some pixels on the bottom but I can't reproduce it.
when I retry it, the test-pac.anbr was fine.
[PacMan-test.zip](https://github.com/user-attachments/files/18260009/PacMan-test.zip)

**Comment by Sleuthgame** [comment]
I still have to test the Flex and Align option, but for the moment I think it's okay for release in January.
I have to write some tutorial about animbrushes and this new tool.
There are a lot of UX and bug fixes done. And It's not a good thing to wait to long between releases as people told us.

Maybe the **Brush Grid** is not exactly as you imagine. But what I see is working nicely. And I don't remember such a tool in any DPaint. It's actually a big step forward  for all users playing around with video-games sprites, tiles and fonts.

Even for anyone wanting to test animbrush, it's a lot easier to figure out what's happening from a still image that to prepare a whole animation to grab a part (I just experience it with my "grab pacman" test.)

**Comment by Sleuthgame** [comment]
I see! With this example from Barbarian 2 (which is not from the Amiga), the inconsistent sprite sizes and the 2-pixel separation between sprites are the most complicated case.

But if I take a more commun tile-based spritesheet like :
https://amiga.lychesis.net/games/BattleSquadron.html

It's working fine !

**Comment by Sleuthgame** [comment]
Ok. My only worries are about ergonomics. Beginners will be lost.
![image](https://github.com/user-attachments/assets/c9fe1725-bdd1-4174-bfd2-c78656c0b8e7)

- add "Brush grid" to the Anim Brush submenu but rename it as "Grab from picture / tile / sprite sheet" or something obvious.

_(I get stuck in my mind of brush grid being an option of grid. If you want to create a new icon for this, create a second tool slot in grid, it will make sense)_

- Add or change the buttons :
1. **Reset** : resets value to default easy number values
- Offset 0 0
- Size 16 16
- Number 3 1

3. **Ok** : close, keep the new values without grabbing. It enable the user to fix some pixels, select the background color and go back again.
4. **Grab** (Pick up is not so obvious for non-english speakers and longuer) : actual Ok behaviour.

**Comment by Sleuthgame** [comment]
Testing **Flex** on this sprite sheet : https://www.spriters-resource.com/fullview/153824/
It's working like a charm. I can't believe it's autodetect. Do it use the background color ?
Note : thoses are not real sprite sheets. Thoses are ripped, it's a recreation. So spacing and disposition may be irrelevant.

**Comment by mriale** [comment]
> Testing **Flex** on this sprite sheet : https://www.spriters-resource.com/fullview/153824/ It's working like a charm. I can't believe it's autodetect. Do it use the current background color @mriale ?

Yes, it uses the background color to detect empty space around a sprite.

**Comment by mriale** [comment]
Great usability ideas!

Added a few things in develop:
- Reset button in Brush Grid Requestor
- Made presets:
  - Offset: 10 20
  - Size: 16 16
  - Number: 3 1
- Added Grab button and made OK save the settings but not grab the brush
- Made the menu items:
  - Grab Frames (alt-b)
  - Grab Grid (alt-g)
- Made all the handles on the grid draggable
- Removed the weird shortcut for brush grid: left-click on brush tool then right-click on brush tool

**Comment by Sleuthgame** [comment]
Great. Can't wait to test all thoses new features !

**Comment by Sleuthgame** [comment]
Isn'it adorable ?

![heli](https://github.com/user-attachments/assets/4442c8ad-ffae-406a-bfa7-a24b47534f64)
![image](https://github.com/user-attachments/assets/f8abd5bf-9dda-42a0-84ef-e1c682f6be37)

**Comment by Sleuthgame** [comment]
I'm sorry but the grabber miss the last line. The logic in DPaint has always been to grab all the pixels in the brush capture area, including its outline. Up line is fine. Last bottom line is forgotten. 😭 
![image](https://github.com/user-attachments/assets/7a2dfc6a-46fd-4903-b7c4-bc43607f287e)
[heli.zip](https://github.com/user-attachments/files/18273438/heli.zip) <= TEST HERE

**Comment by mriale** [comment]
That's a cute helicopter!

The way the grid works is that the top and left side of each grid cell is grabbed and the bottom and right side is the top and left side of adjacent cells. Not sure how to make that clearer unless I make dotted lines like on the flex options. So to grab the helicopter you would have to make it one pixel taller.

**Comment by Sleuthgame** [comment]
Ok, I get it. Nothing a good tutorial can't explain 😉 
Happy New Year's Eve!

**Comment by mriale** [comment]
Added the Animbrush Settings requestor:
![image](https://github.com/user-attachments/assets/85c248be-12d5-4a92-afa7-9dafd85bc3ed)

Of course I also had to add the functionality that makes this work:
- animation direction (forward/backward/ping-pong)
- non-integer frame incrementing (for duration)

Also, I added random frames as a bonus.

**Comment by mriale** [comment]
Yeah, the Random buttons bothered me also, so I lined them up:
![image](https://github.com/user-attachments/assets/5dc769e3-06f1-4cae-aecb-d57362a4b3ce)

**Comment by Sleuthgame** [comment]
👍 
Let's go testing.

**Comment by mriale** [comment]
I think I'm done messing with this. I didn't really like the way the random function worked before. It used to:
- create a list of 4 ranges of the brush cycling up
- randomize the list

You could see patterns after a few times.

Now it just chooses an animbrush randomly every time it changes.

**Comment by Sleuthgame** [comment]
At least, I almost have my bush generator. Dot is too fast, standard Spray too but Circle Airbush 2 is fine.
![image](https://github.com/user-attachments/assets/e9f5a641-142e-47b9-832f-4ecaee9b0265)

**Comment by Sleuthgame** [comment]
Interesting ! Curve with **Airbrush** 1, animbrush Random.
![image](https://github.com/user-attachments/assets/5b935d3a-98be-4024-a48c-e7db8f84ae55)

Can you lower the "flood" speed when using an animbrush with dots, and most drawing tools ?

**Comment by Sleuthgame** [comment]
It began to be a very creative drawing tool.

N total, Every N dot, Airbrush and Continous with a 4 frame leave animbrush.

Even with Airbush at 1, it's too much.
![image](https://github.com/user-attachments/assets/a00acdc8-6c99-4fdc-a854-5c55b3c4f898)

**Comment by Sleuthgame** [comment]
Heve fun !
[ANIMBRUSH.zip](https://github.com/user-attachments/files/18302767/ANIMBRUSH.zip)

**Comment by Sleuthgame** [comment]
I try hard to figure out some animbrushes usefull for drawing but thoses tools are not really designed with that possibility. User can't understand how to use it. It's not really a cloner, not a quite a animated pen, not a randomizer. There are obviously something to dig here, but we have to make it clear and easy to use. I drop that idea and concentrate on sprite and animation features tutorial.

![image](https://github.com/user-attachments/assets/0bd32071-c243-4315-8e5a-860bbe3d873a)
![image](https://github.com/user-attachments/assets/de5faf88-6f2c-4c13-9596-9c8ee35912cd)

**Comment by Sleuthgame** [comment]
![image](https://github.com/user-attachments/assets/af413ba4-860c-49bd-a3ae-e43812409adc)
**WIP**

**Comment by Sleuthgame** [comment]
Here we go !
![PyDPainter](https://github.com/user-attachments/assets/ccd34bc3-a425-419a-88a7-eacd110c0e36)

**Comment by mriale** [comment]
> If I preview the **Curve**, every time I move, it refresh with a different "distribution" (Airbrush 1). Very fine. If I like what I see then click, the distribution is not like the preview, it's another, which is a bit confusing.
> 
> Same with line, rect, circle etc.

Now it doesn't redraw the current tool when finalizing it unless it was interrupted by moving the mouse. That means the last airbrush you see is the one you get. I added this to each of the drawing tools, so it touched a lot of code.

> Even with Airbush at 1, it's too much.

I made the Airbrush on the Spacing requestor fractional so you set it between 0 and 1. I also added spinners to all the values on the Spacing requestor. The spinners on the Airbrush value halve or double the fractional value when it goes below 1.

![image](https://github.com/user-attachments/assets/4898ad3a-2dac-4b5e-a3f7-2e532151ef40)

**Comment by mriale** [comment]
I'm also hoping to add a collapse button on draggable requestors like the Brush Grid requestor:
![image](https://github.com/user-attachments/assets/1b919626-332c-4d97-a4b4-55db8b569596)

to make it do this:
![image](https://github.com/user-attachments/assets/72e75d62-6bac-4e31-99be-55d0863b6002)
so you can see things on the screen instead of it being blocked by the requestor.

I think I could add this easily without much impact.

**Comment by Sleuthgame** [comment]
Greats improvement here ! Kudos !

I try to focus and deliver the animbrush tutorial.

**Comment by Sleuthgame** [comment]
Obviously a crash when right clicking any drawing tool settings. Do you experience it as well or need a crash dump for this ?

Just in case, I remember yesterday, the Cycling mode was no more functional with Animbrushes.
Good luck !

**Comment by mriale** [comment]
Hmmmm... Sorry, it's not obvious to me. I've been right-clicking everywhere and no crashes for me. How do you do it?

Also I tried cycling mode with animbrushes and it works for me. Maybe you have multicycle checked?

**Comment by Sleuthgame** [comment]
> Hmmmm... Sorry, it's not obvious to me. I've been right-clicking everywhere and no crashes for me. How do you do it?

Ok, let me download for sure the last develop and try again, locate the dump and send it to you.
> Also I tried cycling mode with animbrushes and it works for me. Maybe you have multicycle checked?

Yes I have. Gasp. :-|
Need to add this to documentation.

**Comment by Sleuthgame** [comment]
If I right click on Line or Curve or any drwaing tool to open Space settings =>crash and exit.
[2025-01-05_150322.zip](https://github.com/user-attachments/files/18310472/2025-01-05_150322.zip)

**Comment by mriale** [comment]
Hopefully that fixed it. Looks like I used a function that was in a newer version of Python.

**Comment by Sleuthgame** [comment]
It won't launch anymore. So I erase all my dev directory. I run install-pygame.bat. Do I need to update something else ?

**Comment by Sleuthgame** [comment]
We are still clicking on PyDPainter.py, aren't we ? Is there a verbose mode I can send you ?

**Comment by Sleuthgame** [comment]
Installing pygame package...
Requirement already satisfied: pygame in c:\users\utilisateur\appdata\local\programs\python\python311\lib\site-packages (2.6.1)

[notice] A new release of pip available: 22.3.1 -> 24.3.1
[notice] To update, run: C:\Users\Utilisateur\AppData\Local\Programs\Python\Python311\python.exe -m pip install --upgrade pip
Installing numpy package...
Requirement already satisfied: numpy in c:\users\utilisateur\appdata\local\programs\python\python311\lib\site-packages (2.2.1)

[notice] A new release of pip available: 22.3.1 -> 24.3.1
[notice] To update, run: C:\Users\Utilisateur\AppData\Local\Programs\Python\Python311\python.exe -m pip install --upgrade pip
Done.
Appuyez sur une touche pour continuer...

**Comment by Sleuthgame** [comment]
C:\Users\Utilisateur>C:\Users\Utilisateur\AppData\Local\Programs\Python\Python311\python.exe -m pip install --upgrade pip
Requirement already satisfied: pip in c:\users\utilisateur\appdata\local\programs\python\python311\lib\site-packages (22.3.1)
Collecting pip
  Downloading pip-24.3.1-py3-none-any.whl (1.8 MB)
     ---------------------------------------- 1.8/1.8 MB 8.9 MB/s eta 0:00:00
Installing collected packages: pip
  Attempting uninstall: pip
    Found existing installation: pip 22.3.1
    Uninstalling pip-22.3.1:
      Successfully uninstalled pip-22.3.1
  **WARNING**: The scripts pip.exe, pip3.11.exe and pip3.exe are installed in 'C:\Users\Utilisateur\AppData\Local\Programs\Python\Python311\Scripts' which is not on PATH.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
Successfully installed pip-24.3.1

**Comment by Sleuthgame** [comment]
AAARGH !
Even the current version isn't launching anymore and I have this error :
![image](https://github.com/user-attachments/assets/93f994c6-6da8-455f-a2d5-932c8552a10e)

Traceback (most recent call last):
  File "PyDPainter.py", line 118, in <module>
  File "PyDPainter.py", line 105, in check_recover
  File "tkinter\__init__.py", line 4130, in __init__
  File "tkinter\__init__.py", line 4075, in __init__
_tkinter.TclError: couldn't open "C:\Users\Utilisateur\.pydpainter-recover\2025-01-05_150322.zip\recover.png": no such file or directory

**Comment by Sleuthgame** [comment]
Thanks for listening and all that hard work on exciting new parameters. Unfortunatly, my PydPainter copy began to crash and I can't test it, but I will. Today, I manage to make progress on Animbrush tutorial.

> > If I preview the **Curve**, every time I move, it refresh with a different "distribution" (Airbrush 1). Very fine. If I like what I see then click, the distribution is not like the preview, it's another, which is a bit confusing.
> > Same with line, rect, circle etc.
> 
> Now it doesn't redraw the current tool when finalizing it unless it was interrupted by moving the mouse. That means the last airbrush you see is the one you get. I added this to each of the drawing tools, so it touched a lot of code.
> 
> > Even with Airbush at 1, it's too much.
> 
> I made the Airbrush on the Spacing requestor fractional so you set it between 0 and 1. I also added spinners to all the values on the Spacing requestor. The spinners on the Airbrush value halve or double the fractional value when it goes below 1.

**Comment by mriale** [comment]
> Thanks for listening and all that hard work on exciting new parameters. Unfortunatly, my PydPainter copy began to crash and I can't test it, but I will. Today, I manage to make progress on Animbrush tutorial.

Sorry, I'll try reinstalling on Windows and see if I have to redo the install script.

**Comment by mriale** [comment]
>_tkinter.TclError: couldn't open "C:\Users\Utilisateur.pydpainter-recover\2025-01-05_150322.zip\recover.png": no such file or directory

It looks like the zip file you sent me was in the recover directory and that was messing up the PyDPainter recovery process. I made it more robust so it can ignore extraneous files in the recover directory now. Sorry about that!

You should be able to either delete the `.pydpainter-recover` directory or get the newest `develop` version for it to work again.

Also on Windows 10, I just uninstalled Python and deleted the Python libraries. After I installed the latest version of Python and ran the `install_pygame.bat` script, PyDPainter launched right away. At least for me, the scripts still work in Windows.

**Comment by mriale** [comment]
Added a `-`/`+` gadget to draggable requestors to act like a roll-up window shade:
![image](https://github.com/user-attachments/assets/ca56a039-c451-4167-934f-ab410c28f42c)

These are all the draggable requestors so they have the new button:
- Stencil
- Palette
- Brush Grid

**Comment by Sleuthgame** [comment]
This disco nightclub is crowdy 😆 
![Dancing-girls2](https://github.com/user-attachments/assets/b1a052db-4211-45c6-be04-dd783d5275d3)

Joke part, the pixel marks at the bottom wasn't in the original animbrush. It's not the first time I reload a saved animbrush and some noisy pixels were added (always from foreground color).
It's hard to reproduce, but it happens.
- save a animbrush
- change screen size or New project ?
- reload animbrush. It's lightly corrupted by unwanted pixels.

**Comment by Sleuthgame** [comment]
Creepy one !
I load again my last ANIM save of this morning. It seems PydPainter take notice for my last **Grab Frames** brush, at the frame 268 itself (I grab the last one with 3 dots to shape exactly all the dancer moves) ...and save it ? (4 colors, 268 frames big project) :
![image](https://github.com/user-attachments/assets/5dad1b06-764a-466c-894a-00d7a9a55d6e)

To be clear, I never drawn this dotted rectangle around my grabbing zone. But it's exactly my grabbing selection.
Whole ANIM + corrupted ANIMBRUSH -> [Dancing.zip](https://github.com/user-attachments/files/18321291/Dancing.zip)

**Comment by Sleuthgame** [comment]
> Added a `-`/`+` gadget to draggable requestors to act like a roll-up window shade:
> * Stencil
> * Palette
> * Brush Grid

Seems robust. Have tried Pick color with Palette and Stencil.

**Comment by Sleuthgame** [comment]
**Pattern fill** is almost working with animbrush. It just reset to the first frame each time you fill. But 7/8 key is working, so we can browse through frames, which is amazing.

![image](https://github.com/user-attachments/assets/7ee79c0b-f9f7-4e73-a99b-c2c3496e8a38)

**Comment by mriale** [comment]
> - reload animbrush. It's lightly corrupted by unwanted pixels.

I found a way to reuse the existing ANIM compression code instead of new code I wrote just for animbrushes. I think this resolved the issue and as a bonus made the file sizes smaller.

I also now clear the dotted rectangle before creating the animbrush. I was able to grab the animbrush from your animation without issues after the fix.

**Comment by Sleuthgame** [comment]
–  Sound good. Have you tried to load it again after a while ? It works on the same session, then it appears corrupt only after a while.

Anyway, more users/testers will tell us...

– When do you plan to release ? I need next weekend to finish the tutorial.

– Will be nice to have some animbrush samples for user to play with. I use the standard palette for my heli and star. Can you send you own animbrushs shapes ?

**Comment by mriale** [comment]
I'm not sure how to reproduce the animbrush corruption problem. Played with it some after I put my fix in. Maybe I can add event recording/playback for a debug mode, but that seems a little involved. Or maybe my fix worked?

I'd like to release this month (January 2025) but if things are not ready, I **don't** want to push it out anyway.

This animbrush is my favorite to play with and easy to make:
![circle](https://github.com/user-attachments/assets/86d79e07-5086-467e-9549-9a83f3698160)

I use a spacing of Every nth dot = 3. You can also make the brush ping-pong for different effects.

I've also used this boring brush for testing:
![6](https://github.com/user-attachments/assets/aae9b3a2-1f27-48f4-a7f3-5359319570b5)

I haven't made any real artistic animbrushes.

**Comment by Sleuthgame** [comment]
> I'm not sure how to reproduce the animbrush corruption problem. Played with it some after I put my fix in. Maybe I can add event recording/playback for a debug mode, but that seems a little involved. Or maybe my fix worked?

I didn't test since your fix, but I will.
 
> I'd like to release this month (January 2025) but if things are not ready, I **don't** want to push it out anyway.
okay, good timing. I've got a bit of time on my hands, but I always tend to motivate myself more on PydPainter.

> This animbrush is my favorite to play with and easy to make: ![circle] 
> I use a spacing of Every nth dot = 3. You can also make the brush ping-pong for different effects.

Oh, that's great !
 
> I've also used this boring brush for testing: [countdown]

A Matrix-like weird characters fall will be nice. I'll try.

> I haven't made any real artistic animbrushes.

I did not really have experience on that, so any suggestion is welcome. I have 4 on my own... I'm working on a flapping bird video conversion...

**Comment by Sleuthgame** [comment]
This is a nice playable brush, Thanks for sharing !
![image](https://github.com/user-attachments/assets/ed687e69-7c87-43ba-8de0-6be8646fdada)
